### PR TITLE
Migrate to ruff format

### LINKED ISSUE #3477: Switch to `ruff format`
Ruff can now format files!

Would mostly require changes in `pyproject.toml` and `.pre-commit-config.yaml`

### PR DESCRIPTION
**Pull request by Saransh-cpp**
# Description

Sorry for creating this huge PR just before the holidays 😬 

Fixes #3477 
Also addresses #3489

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [ ] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by agriyakhetarpal** [comment]
Should we close https://github.com/pybamm-team/PyBaMM/pull/3492?

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -9,12 +9,14 @@ repos:
...
         types_or: [python, pyi, jupyter]
+      - id: ruff-format
+        types_or: [python, pyi, jupyter]
 
   - repo: https://github.com/adamchainz/blacken-docs
     rev: "1.16.0"
     hooks:
        - id: blacken-docs
-         additional_dependencies: [black==22.12.0]
+         additional_dependencies: [black==23.*]
```
Comment (agriyakhetarpal) [inline .pre-commit-config.yaml:19]: I recently got to know that `ruff` is now capable of formatting code blocks in documentation files (Markdown, reST, etc.) as well, should we consider removing `blacken-docs`? Feel free to convert this comment to an issue if yes

**Review by agriyakhetarpal — APPROVED** [review]
Thanks, @Saransh-cpp! Looks good and it is nice to get this tool working for notebooks. Is there a way to reduce the line length in the configuration just for the notebooks? See [this notebook in the PR build](https://pybamm--3656.org.readthedocs.build/en/3656/source/examples/notebooks/creating_models/2-a-pde-model.html#Defining-a-geometry-and-mesh) for example – it would be great if we can set the line length to, say, 60 for notebooks, while keeping it at 88 or so for Python files. It reduces the amount of side scrolling required when browsing through the code cells for accessibility reasons.

In hindsight, we should have kept #3492 open since the contributor was active for a while and it had been just three weeks since they opened it. Let's be more patient next time we encounter a delay and wait for at least two months, given that this was not a high priority issue to fix.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (Saransh-cpp) [reply .pre-commit-config.yaml:19]: I don't think ruff can do that yet, see - https://github.com/astral-sh/ruff/issues/8237 and https://github.com/astral-sh/ruff/issues/3792

**Comment by Saransh-cpp** [comment]
> it would be great if we can set the line length to, say, 60 for notebooks, while keeping it at 88 or so for Python files

I went through ruff's config and I don't think they support such flexibility at the moment.

**Comment by agriyakhetarpal** [comment]
How about temporarily setting it to 60 and linting the notebooks while ignoring the rest, then setting it back to the default and linting the rest ignoring the notebooks?

**Comment by Saransh-cpp** [comment]
> How about temporarily setting it to 60 and linting the notebooks while ignoring the rest, then setting it back to the default and linting the rest ignoring the notebooks?

ruff will automatically expand the lines with length = 60 to length = 88 whenever we run it again (that is, in the CI), it goes both ways - shrink as well as expand line lengths

**Comment by agriyakhetarpal** [comment]
Ah I didn't know that, I will open up a feature request about this upstream and hope they find it reasonable

**Pull request merged**